ls1intum/Artemis
# `General`: Improve the responsiveness of the navigation bar for smaller devices

### Checklist
#### General


#### Client

### Motivation and Context



The navbar has a fixed breaking point for collapsing everything into a menu. This leads to a lot of unused space on certain screen sizes. In the example below, everything would still easily fit into the navbar itself.

![grafik](https://user-images.githubusercontent.com/23171488/178299066-f72dd14a-c196-4b23-91c0-162ff0152b3d.png)

### Description


This PR adds some more advanced responsive behavior to the navbar. If there's not enough space, the links to Course Overview, Course Management and Server Administration are collapsed into the menu, but the icons for notifications, theme and the user menu stay in the menu bar. Then, if there's not enough space for them, they'll be moved into the collapsed menu in a second step.
The breakpoints are approximately and dynamically calculated, taking the length of the login and the available options into account. For example, the breakpoint for the first wrap step needs to be lower if someone doesn't have access to Server Administration.

### Steps for Testing

Prerequisites:
- A user of each type: Student, Instructor, Admin

Test that the navbar collapses correctly for each user type. You should observe different breakpoints for each as they have different navbar options available. Note that the navbar will always leave space for the maximum occupancy (notification badge, disconnected icon, loading spinner) so it might look like it's wrapping too early but it's actually not - and it's not the intention to be 100% precise. It should just not let 1000 pixels of wasted space, but be a bit more intelligent

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| navbar.component.ts | 76% | 86% |

### Screenshots




https://user-images.githubusercontent.com/23171488/178323269-1be0eeb0-128e-479b-850f-a7a7ae20a6f1.mov
